[huixuant] iP

huixuant opened the pull request:
### Peppa 

> "Oink" means "I love you" in piggy language! – Unknown

Peppa is your trustable right-hand man (or piggy? 🐷). It frees your mind of having to remember things you need to do. It's 

* text-based
* easy and fast to use
* ~~cute~~ _SUPER CUTE!!!_

All you need to do is:
1. [Download](https://github.com/huixuant/ip/releases/download/A-Jar/peppa-v0.1.jar) it
2. Follow the instructions [here](https://github.com/huixuant/ip/releases/tag/A-Jar)
3. Add your tasks
4. Sit back, relax and let it manage your tasks for you :)

Most importantly, it's **FREE**!

Features:
- [x] Managing tasks (todos, deadlines, events)
- [ ] Reminders (coming soon!)

View a snippet of the `main` method here:
```java
public static void main(String[] args) {
    new Peppa(FILE_PATH).run();
}
```

Commit:
Add Gradle support

huixuant committed:
Implemented Level-1 (greet, echo, exit) and personalised chatbot

huixuant committed:
Implemented Level-2 (add, list)

huixuant committed:
Implemented Level-3 (mark as done)

huixuant committed:
Implemented Level-4 (todos, events, deadlines)

huixuant committed:
Added test cases and expected outputs for automated text UI testing

huixuant committed:
Refactored code to handle exceptions like incorrect input

huixuant committed:
Implemented Level-6 (delete)

huixuant committed:
Refactored code to implement enums

huixuant committed:
Implement Level-7 (add support for file operations)

huixuant committed:
Implement Level-8 (add support for date and time)

huixuant committed:
Merge branch 'branch-Level-7'

huixuant committed:
Merge branch 'branch-Level-8'

huixuant committed:
Partially implement A-MoreOOP by abstracting user interactions into Ui class

huixuant committed:
Implement A-MoreOOP (add Ui, Storage, Parser, Tasklist and Command classes)

huixuant committed:
Merge branch 'add-gradle-support' of github.com:huixuant/ip

huixuant committed:
Implement A-JUnit (add tests for selected methods)

huixuant committed:
Add Javadoc header comments

huixuant committed:
Edit code to comply with coding standard

huixuant committed:
Add find task with keyword functionality

huixuant committed:
Merge branch 'branch-A-JavaDoc'

huixuant committed:
Merge branch 'branch-A-CodingStandard'

huixuant committed:
Merge branch 'Level-9'

mmaimer33 commented on src/main/java/peppa/Deadline.java:
```diff
@@ -0,0 +1,28 @@
+ */
+public class Deadline extends Task {
+    private LocalDateTime dueDate;
+
```
Would recommend adding JavaDoc comments for constructors.

mmaimer33 commented on src/main/java/peppa/Deadline.java:
```diff
@@ -0,0 +1,28 @@
+    @Override
+    public String toString() {
+        return "[D]" + super.toString() + " (by: "
+                + this.dueDate.format(DateTimeFormatter.ofPattern("MMM dd yyyy',' hh'.'mma")) + ")";
```
Nice use of line breaks for readability :).

mmaimer33 commented on src/main/java/peppa/Event.java:
```diff
@@ -0,0 +1,32 @@
+    private LocalDate from;
+    private LocalDate to;
+
+    public Event(String name, LocalDate from, LocalDate to) {
```
Similar JavaDoc recommendation for this and other constructors.

mmaimer33 commented on src/main/java/peppa/Parser.java:
```diff
@@ -0,0 +1,219 @@
+    public static Command parseCommand(String fullCommand) {
+        String[] args = fullCommand.split(" ");
+        String commandType = args[0];
+        switch (commandType) {
+        case ListCommand.COMMAND_WORD:
+            return new ListCommand();
+        case MarkCommand.COMMAND_WORD:
+            return parseMarkCommand(args);
+        case UnmarkCommand.COMMAND_WORD:
+            return parseUnmarkCommand(args);
+        case DeleteCommand.COMMAND_WORD:
+            return parseDeleteCommand(args);
+        case FindCommand.COMMAND_WORD:
+            return new FindCommand(args[1]);
+        case TodoCommand.COMMAND_WORD:
+            return parseTodoCommand(fullCommand);
+        case DeadlineCommand.COMMAND_WORD:
+            return parseDeadlineCommand(fullCommand);
+        case EventCommand.COMMAND_WORD:
+            return parseEventCommand(fullCommand);
+        case ExitCommand.COMMAND_WORD:
+            return new ExitCommand();
+        default:
+            return new InvalidCommand();
```
Well done avoiding the common error to indent case statements further.

mmaimer33 commented on src/main/java/peppa/Task.java:
```diff
@@ -0,0 +1,34 @@
+
+public abstract class Task {
+    protected String name;
+    protected boolean done;
```
```suggestion
    protected boolean isDone;
```
(Acc. to boolean naming conventions)

mmaimer33 commented on src/main/java/peppa/TaskList.java:
```diff
@@ -0,0 +1,60 @@
+        taskList.remove(task);
+    }
+
+    public ArrayList<Task> findTasks(String keyword) {
```
Possible missing JavaDoc comment?

mmaimer33 commented on src/main/java/peppa/commands/DeadlineCommand.java:
```diff
@@ -0,0 +1,76 @@
+        }
+    }
+
+    @Override
```
Maybe you could add JavaDoc comments for the overrides since they all have different implementations? Alternatively just one in the interface to be inherited by the subclasses?

mmaimer33 commented on src/test/java/peppa/DeadlineCommandTest.java:
```diff
@@ -0,0 +1,18 @@
+
+public class DeadlineCommandTest {
+    @Test
+    public void getParameterValue_noDescription_exceptionThrown() {
```
Good name for a test 👍🏻

mmaimer33 reviewed (commented):
Looks great! Coding standard mostly followed, apart from some nits.

huixuant committed:
Implement GUI using JavaFX

huixuant committed:
Merge branch 'branch-Level-10'

huixuant committed:
Add in missing JavaDoc comments (excluding controllers)

huixuant committed:
Modify find command to accept multiple keywords using varargs

MrTwit99 commented on src/main/java/peppa/Peppa.java:
```diff
@@ -0,0 +1,56 @@
+ * Represents the chatbot application.
+ */
+public class Peppa {
+    public static final String FILE_PATH = "data/todo.txt";
```
I like how you made the file path a constant name so as to promote encapsulation. 
I would incorporate this idea as well!

Just one quick note, since you are declaring FILE_PATH and using in the same class, would it be better to give it private instead of public as modifier?

```suggestion
    private static final String FILE_PATH = "data/todo.txt";
```

MrTwit99 commented on src/main/java/peppa/Storage.java:
```diff
@@ -0,0 +1,71 @@
+import java.util.Scanner;
+
+/**
+ * Represents a file for storing tasks and handles all file I/O-related logic.
```
Optional: If you would like to, you could add @author as part of the JavaDocs in every class so as to claim "copyrights". 

Note: Ps this is not part of the Java Coding Standards but I just feel since you have made this pretty unique and put in a lot of hardwork, might be good to call dips on "copyright".

MrTwit99 commented on src/main/java/peppa/TaskList.java:
```diff
@@ -0,0 +1,75 @@
+     * @param keywords Keywords that the user is searching for.
+     * @return Subset of tasks in the master list that match the given keyword(s).
+     */
+    public HashMap<String, ArrayList<Task>> findTasks(String[] keywords) {
+        HashMap<String, ArrayList<Task>> map = new HashMap<>();
+        for (String keyword : keywords) {
+            ArrayList<Task> matchedTasks = new ArrayList<>();
+            for (Task task : taskList) {
+                if (task.name.contains(keyword)) {
+                    matchedTasks.add(task);
+                }
+            }
+            map.put(keyword, matchedTasks);
+        }
+        return map;
+    }
```
I like your idea of allowing the findTasks() to be able to search for words given in "find xxxx xxxx" individually instead of a substring so as to get better search results!

MrTwit99 commented on src/main/java/peppa/Storage.java:
```diff
@@ -0,0 +1,71 @@
+
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.util.Scanner;
```
I'm not sure if its better or part of coding standards but iirc in 2030S we used to leave an empty line for imports of different directories. I think it might be a good idea you could consider.

```suggestion
import java.io.FileWriter;
import java.io.IOException;

import java.util.Scanner;

```

MrTwit99 commented on src/main/java/peppa/Ui.java:
```diff
@@ -0,0 +1,185 @@
+     * Prints the introductory greeting message upon initialisation of the chatbot.
+     */
+    public static void greetUser() {
+        String logo = " ____  ____  ____  ____   __\n"
+                + "(  _ \\(  __)(  _ \\(  _ \\ / _\\\n"
+                + " ) __/ ) _)  ) __/ ) __//    \\\n"
+                + "(__)  (____)(__)  (__)  \\_/\\_/\n";
+        String art = "       _\n"
+                + "  <`--'\\>______\n"
+                + "  /. .  `'     \\\n"
+                + " (`')  ,        @\n"
+                + "  `-._,        /\n"
+                + "     )-)_/--( >  jv\n"
+                + "    ''''  ''''\n";
```
Nice icon you got there!

MrTwit99 commented on src/main/java/peppa/Parser.java:
```diff
@@ -0,0 +1,231 @@
+import peppa.commands.EventCommand;
+import peppa.commands.ExitCommand;
+import peppa.commands.FindCommand;
+import peppa.commands.IncorrectCommand;
+import peppa.commands.InvalidCommand;
```
Would it be better to make IncorrectCommand and InvalidCommand exceptions too under PeppaExceptions and not as commands?  

There wouldn't be much changes to existing IncorrectCommand and InvalidCommand java files since it prints its error message only iirc.

MrTwit99 reviewed (commented):
Overall, I like how you did your Peppa application. It's nice and easy to follow through the code and some of the things were pretty unique like the displayed messages etc. It's pretty amazing how you even created an interface (or sth) for it (aka Launcher.java, Main.java etc) and followed the OOP principles and Coding standards.

There are a couple suggestions below which you could look through and see. Not much of an issue overall and these suggestions is optional.

kenzantonius reviewed (commented):
Great job! I do not find anything wrong in the Code Quality. Keep up the good work!

One thing to note: There are some parts that you did not give Java Comments.

huixuant committed:
Add Java assertions

huixuant committed:
Improve code quality

huixuant committed:
Merge pull request #1 from huixuant/branch-A-Assertions

Add Java assertions

huixuant committed:
Merge branch 'master' into branch-A-CodeQuality

huixuant committed:
Merge pull request #2 from huixuant/branch-A-CodeQuality

Refactor code to improve code quality

huixuant committed:
Set up continuous integration using Github Actions

huixuant committed:
Merge branch 'master' of github.com:huixuant/ip

huixuant committed:
Resize height of DialogBox to fit content

DialogBox truncates longer messages with an ellipsis.

Displaying the full message content is more user-friendly.

Let's,
* update the fxml file to set minHeight of Label to USE_PREF_SIZE

huixuant committed:
Implement FlexibleDataSource feature

Peppa always loads task data from data/todo.txt, if it exists.

Allowing users to choose an alternative file to load tasks from provides
greater flexibility and optimises the user experience.

Let's create new commands (FilesCommand and SelectCommand) to support
viewing of current data sources and switching of data source.

huixuant committed:
Merge pull request #3 from huixuant/branch-BCD-Extension

Implement FlexibleDataSource feature

huixuant committed:
Make improvements to GUI

huixuant committed:
Correct xmlns field value in FXML files

huixuant committed:
Improve GUI to allow resizing of window

huixuant committed:
Add screenshot of Ui and update README

huixuant committed:
Fix Checkstyle violations and ExitCommand behaviour
